core/src/apps/bitcoin/verification.py

```diff
@@ -0,0 +1,104 @@
+                    raise wire.DataError("Invalid script hash")
+                self.public_keys, self.threshold = parse_output_script_multisig(script)
+            else:
+                raise wire.DataError("Unsupported signature script")
+        elif witness and witness != b"\x00":
+            if len(script_sig) == 23:  # P2WPKH nested in BIP16 P2SH
+                public_key, signature, hash_type = parse_witness_p2wpkh(witness)
+                pubkey_hash = ecdsa_hash_pubkey(public_key, coin)
+                if input_script_p2wpkh_in_p2sh(pubkey_hash) != script_sig:
+                    raise wire.DataError("Invalid public key hash")
+                script_hash = coin.script_hash(script_sig[1:])
+                if output_script_p2sh(script_hash) != script_pubkey:
+                    raise wire.DataError("Invalid script hash")
+                self.public_keys = [public_key]
+                self.signatures = [(signature, hash_type)]
+            elif len(script_sig) == 35:  # P2WSH nested in BIP16 P2SH
+                script, self.signatures = parse_witness_p2wsh(witness)
+                script_hash = sha256(script).digest()
+                if input_script_p2wsh_in_p2sh(script_hash) != script_sig:
+                    raise wire.DataError("Invalid script hash")
+                script_hash = coin.script_hash(script_sig[1:])
+                if output_script_p2sh(script_hash) != script_pubkey:
+                    raise wire.DataError("Invalid script hash")
+                self.public_keys, self.threshold = parse_output_script_multisig(script)
+            else:
+                raise wire.DataError("Unsupported signature script")
+        else:
+            if len(script_pubkey) == 25:  # P2PKH
+                public_key, signature, hash_type = parse_input_script_p2pkh(script_sig)
+                pubkey_hash = ecdsa_hash_pubkey(public_key, coin)
+                if output_script_p2pkh(pubkey_hash) != script_pubkey:
+                    raise wire.DataError("Invalid public key hash")
+                self.public_keys = [public_key]
+                self.signatures = [(signature, hash_type)]
+            elif len(script_pubkey) == 23:  # P2SH
+                script, self.signatures = parse_input_script_multisig(script_sig)
+                script_hash = coin.script_hash(script)
+                if output_script_p2sh(script_hash) != script_pubkey:
+                    raise wire.DataError("Invalid script hash")
+                self.public_keys, self.threshold = parse_output_script_multisig(script)
+            else:
+                raise wire.DataError("Unsupported signature script")
+
+        if self.threshold != len(self.signatures):
+            raise wire.DataError("Invalid signature")
+
+    def ensure_hash_type(self, hash_type: int) -> None:
+        if any(h != hash_type for _, h in self.signatures):
+            raise wire.DataError("Unsupported sighash type")
+
+    def verify(self, digest: bytes) -> None:
```

Comment:
I think that such an expectation is unfounded and `ensure_verify` sounds plain confusing. I guess I don't mind returning bool, raising an exception just makes the code a little cleaner.

Reply:
FWIW there are inconsitencies in our codebase wrt `verify()` returning or raising.

My preference would be for raising, but consistency is more important. This might warrant a separate issue.

Reply:
I beg your pardon, the expectation is based on my experiences, I can't remember I've ever encountered a function that is called `verify`, verifies a signature and throws an exception if the signature is not valid.

I'm not against a function that verifies a signature and throws an exception if the signature is not valid. I'm against such the function that is called `verify`. 

The reason (as I said) is that I expect that such the function returns `bool`. For example [this line](https://github.com/trezor/trezor-firmware/blob/dce49d9d086db36392b45603238b8e84eb851145/core/src/apps/bitcoin/sign_tx/bitcoin.py#L193) at first glance looks like you forgot to check the return value. This habit leads to mistakes such as calling `secp256k1.verify` without checking the return value. That's my opinion.

Reply:
> I beg your pardon, the expectation is based on my experiences, I can't remember I've ever encountered a function that is called `verify`, verifies a signature and throws an exception if the signature is not valid.

Grep our codebase for `verify` and you will find several functions that return `bool` and several that raise exceptions.

It's definitely a good idea to unify this, but it's not unified today and you need to look at the function signature when reading code. Hence if we want to change this, we should first document the expected signature, and make an issue for the non-conforming half or the functions.